### PR TITLE
Add primspec for Sub, Index, Chunk, and Embedding

### PR DESCRIPTION
**Pull request by houseroad**
Add primspec for Sub, Index, Chunk, and Embedding. These ops are used by word language model.

### CODE REVIEW

File: torch/autograd/_functions/tensor.py [hunk 1]
```diff
@@ -10,6 +10,28 @@
 class Index(Function):
 
     @staticmethod
+    def primspec(g, i, index):
+        # We should only expect index as an integer in this case.
+        # We use "Slice" to get the index-th element in i,
+        # Then we reduce the dimension using "Reshape".
+        if not isinstance(index, int):
+            raise ValueError('Right now, only int-type index is suppported.')
+        starts = [0] * len(i.type().sizes())
```
Comment (ezyang) [inline torch/autograd/_functions/tensor.py:19]: Not for this PR, but ToffeeIR master has the slices now as attributes, not tensors, so we'll have to adjust.

File: torch/nn/_functions/thnn/sparse.py [hunk 2]
```diff
@@ -9,6 +9,12 @@
 class Embedding(Function):
 
     @staticmethod
+    def primspec(g, indices, weight, padding_idx, max_norm, norm_type, scale_grad_by_freq,
+                 sparse=False):
+        # TODO More checks on the parameters.
```
Comment (ezyang) [inline torch/nn/_functions/thnn/sparse.py]: Yeah, I'd really prefer it if we checked the parameters.

**Comment by ezyang** [comment]
Modulo the missing param checks in Embedding, looks good.

File: torch/autograd/_functions/tensor.py [hunk 3]
```diff
@@ -576,6 +597,21 @@ def forward(ctx, input, k, dim=None, largest=True, sort=True, return_indices=Tru
...
     @staticmethod
+    def primspec(g, i, num_chunks, dim=0):
+        dim_size = i.type().sizes()[dim]
+        split_size = (dim_size + num_chunks - 1) // num_chunks
+        lengths = []
+        while (dim_size > 0):
+            this_split_size = split_size if dim_size >= split_size else dim_size
+            lengths.append(this_split_size)
+            dim_size = dim_size - split_size
+        result = []
```
Comment (ezyang) [inline torch/autograd/_functions/tensor.py:608]: The lengths calculation seems to be buggy. Repro with this in caffe2_pytorch_test_models.py:

```
    def test_chunk(self):
        class MyModel(torch.nn.Module):
            def __init__(self):
                super(MyModel, self).__init__()

            def forward(self, input):
                return input.chunk(20, dim=2)[0]
        self.run_model_test(MyModel(), train=False, batch_size=BATCH_SIZE)
```

File: torch/nn/_functions/thnn/sparse.py [hunk 4]
```diff
@@ -9,6 +9,27 @@
 class Embedding(Function):
 
     @staticmethod
+    def primspec(g, indices, weight, padding_idx, max_norm, norm_type, scale_grad_by_freq,
+                 sparse=False):
+        if max_norm is not None:
+            raise ValueError('Right now, re-norm is not supported.')
+
+        # To support padding_idx, we set the corresponding element to 0 in weight.
+        if padding_idx is not None:
```
Comment (ezyang) [inline torch/nn/_functions/thnn/sparse.py:18]: This logic is not right, although the existing code in PyTorch seems to take part of the blame.

The crux of the matter is that PyTorch actually *never* passes `padding_idx` as `None` to the autograd function:

```
    def forward(self, input):
        padding_idx = self.padding_idx
        if padding_idx is None:
            padding_idx = -1
        return self._backend.Embedding.apply(
            input, self.weight,
            padding_idx, self.max_norm, self.norm_type,
            self.scale_grad_by_freq, self.sparse
        )
```

padding_idx is passed as -1. But in this case, we SHOULD NOT mask the last element (as we should when `padding_idx=-1`, since the weight in this case is NOT zeroed (as you would expect):

```
    def reset_parameters(self):
        self.weight.data.normal_(0, 1)
        if self.padding_idx is not None:
            self.weight.data[self.padding_idx].fill_(0)
```

I think an easy fix is to replace the test for padding_idx being None with a test if it is -1.

Comment (ezyang) [reply torch/nn/_functions/thnn/sparse.py:18]: I put the fix in https://github.com/ezyang/pytorch/pull/190

Comment (houseroad) [reply torch/nn/_functions/thnn/sparse.py:18]: It's a good idea to write unit test for each primspec. I will double check the logic, and add the test if not existing.

File: torch/autograd/_functions/tensor.py (hunk 3 above)
Comment (houseroad) [reply torch/autograd/_functions/tensor.py:608]: The bug got fixed... If we do chunk(a, 20) in Pytorch, a is a 224 dimension tensor, then the result will only contain 19 elements...

File: torch/nn/_functions/thnn/sparse.py (hunk 4 above)
Comment (houseroad) [reply torch/nn/_functions/thnn/sparse.py:18]: Since the weight won't get changed except in the initializer, we can safely remove the padding logic. In the initializer, the corresponding padding column is already set to 0. I push my fix in #190

Comment (ezyang) [reply torch/nn/_functions/thnn/sparse.py:18]: OK, I guess we don't need the padding logic. BTW, it would be good if we can put word language model in the E2E tests.

**Comment by ezyang** [comment]
Obsoleted by #190
